### ISSUE TITLE
Performance degradation when using combine_reducer and state consists of objects that implement `__eq__`

### ISSUE THREAD

**Issue by zpfvo** [benjamin9999/pydux#17]
The combine function is causing a huge performance degradation by using the equality operator `!=` instead of `is not`.
https://github.com/usrlocalben/pydux/blob/943ca1c75357b9289f55f17ff2d997a66a3313a4/pydux/combine_reducers.py#L77-L78

This causes the whole state to be compared on each action that reaches the reducer, even if the actions are not handled in any reducer at all.

When the state consists of dictionaries, is not that obvious because dictionary comparison is very fast.
When the state consists of 'normal' class objects its also fast because `__eq__` falls back to comparing identity using `is`.

When we modeled our state using pydantic it converts the objects to dicts first to compare for equality when using `!=` which caused a significant performance impact.

We simply fixed it using `is  not` in the combine function.

https://github.com/zuhause-plattform/pydux/commit/f5e2bce948e296868015795d0ec67b5dbdee73b4

**Comment by usrlocalben** [comment]
If a reducer gives a new object instance (dict or otherwise) but has the same _value_ this should fail. i.e. `foo is not bar`, yet `foo == bar`.

You are right, that reducers() that don't handle an action generally follow the pattern of returning the input (reference) so those results would pass the _is_ test. However, it's not a requirement that reducers do this.

The following expr will short-circuit such that you get the _is_ optimization you want without breaking the general value-semantic case.

`has_changed = has_changed or not (next_state_for_key is previous_state_for_key or next_state_for_key == previous_state_for_key)`

```
In [62]: xa = {'foo': 'bar'}
In [63]: xb = xa
In [64]: not (xa is xb or xa == xb)
Out[64]: False

In [65]: xa = {'foo': 'bar'}
In [66]: xb = {'foo': 'bar'}
In [67]: xa is xb
Out[67]: False
In [68]: not (xa is xb or xa == xb)
Out[68]: False
```

**Comment by zpfvo** [comment]
> You are right, that reducers() that don't handle an action generally follow the pattern of returning the input (reference) so those results would pass the is test. However, it's not a requirement that reducers do this.

I didn't know that it was expected behavior to get the old reference back if a new state was returned but the attributes had the same values. IMHO its a little bit counter-intuitive, especially because that behavior is only introduced by the combine_reducer function. The behavior is not present when using only one reducer, or am I missing something?

Anyway I like your optimization. This will negate most of the negative performance impact.
And thanks for your quick response 👍

**Comment by zpfvo** [comment]
The topic kept being stuck in my head, because I couldn't find out the reasoning behind this behavior.
I investigated a little bit more and now I'm convinced that comparing for identity is actually the intended behavior of the original redux implementation in js and using the equality operator might be a mistake that happened while porting to python.

Looking at the implemetation of the [reduxjs repo](https://github.com/reduxjs/redux/blob/9be553a2c5ba04d9d769f31e25a6dc93deb334cf/src/combineReducers.ts#L194-L195) the key line is

`hasChanged = hasChanged || nextStateForKey !== previousStateForKey`

if we test your example above in a javascript environment using `!==` we can see, that there is no equality of the two objects even if the attributes are the same.
```js
❯ node
> xa = {'foo': 'bar'}
{ foo: 'bar' }
> xb = {'foo': 'bar'}
{ foo: 'bar' }
> xa !== xb
true
```

AFAIK there is no way to implement an equality function like we can with `__eq__` in python, so an objects is only equal to itself in js.

_(Disclaimer: I might be wrong Im not a js programmer, thats just what I found out)_

**Comment by usrlocalben** [comment]
You have persuaded me.
Do you want to bring a PR?

**Comment by zpfvo** [comment]
Will happily prepare a PR later.
One thing that I want to note, is that, there are unfortunately some cases where using `is` won't yield the same results as in JS:

```js
❯ node
> a = 10e10
100000000000
> b = 10e10
100000000000
> a === b
true
>
```
while

```python
❯ python3                                                                      
>>> a = 10e10
>>> b = 10e10
>>> a == b
True
>>> a is b
False
>>>
```
I think its a kind of rare case, because most people will use objects to hold state, but if you would just use primitive number than it will break with numbers > 256

```python
>>> a = 256
>>> b = 256
>>> a is b
True
>>> b = 257
>>> a = 257
>>> a is b
False
>>> 

```

**Comment by usrlocalben** [comment]
Interesting.
In the case that anyone out there is using this lib and depends (perhaps unknowingly) on the `==` behavior, it could still be preferred to implement it as I suggested above, using `a is b or a == b`. Do you see any downside?

**Comment by zpfvo** [comment]
Yeah i guess that would be the simplest solution. One could try to exactly mimic the JS comparison, but then you had to do some `isinstance` checks. Don't know if that is feasible.

Another option would be to add an additional optional keyword argument to the `combine_reducers` function, like `compare_identity_only` or something like that, but I don't know if you would consider this elegant.
